### DISCUSSION TITLE
GLSL fragment shader file recognized as JavaScript

### DISCUSSION THREAD

**Discussion by teamclouday** [github-linguist/linguist#5480] [Classification] [answered]
I just found that the language analysis on my [repo](https://github.com/teamclouday/Render/search?l=JavaScript) is incorrectly reporting `.frag` files as `JavaScript` files, where they should be in the category of `GLSL`.

So I searched in this repository and found that `.frag` is defined both in [GLSL](https://github.com/github/linguist/blob/c492ac0d5bc697033c4261e8bfb597949ea882e2/lib/linguist/languages.yml#L1842) and [JavaScript](https://github.com/github/linguist/blob/c492ac0d5bc697033c4261e8bfb597949ea882e2/lib/linguist/languages.yml#L2735).

I'm not an expert in JavaScript, so I'm wondering does `.frag` extension exist in JavaScript programming? Or is it part of WebGL and is essentially GLSL?

Interestingly, this looks like a random error, because the same problem does not occur in [this repo](https://github.com/lettier/3d-game-shaders-for-beginners/search?l=GLSL), but exists in [that one](https://github.com/Ravbug/ShaderTranspiler/search?l=JavaScript), where `.frag` is GLSL in both cases.

Should I open an issue on this?

**Comment by lildude** [comment, marked as answer]
> I'm not an expert in JavaScript, so I'm wondering does .frag extension exist in JavaScript programming

It does… well, in a subset of JavaScript. It seems Angular uses this extension.

> Interestingly, this looks like a random error, because the same problem does not occur in [this repo](https://github.com/lettier/3d-game-shaders-for-beginners/search?l=GLSL), but exists in [that one](https://github.com/Ravbug/ShaderTranspiler/search?l=JavaScript), where `.frag` is GLSL in both cases.

As both languages use the same extension, we need to rely on heuristics and the classifier to determine the language. There isn't currently a heuristic so we rely entirely on the classifier which is trained on the samples we have. In both cases, the samples we have appear to be quite small and similar to my naive eye, so the classifier has a higher chance of getting it wrong.

> Should I open an issue on this?

A PR to improve things would be preferred. Adding a heuristic and more representative samples of each language would help improve things.

**Reply by teamclouday** [reply]
Thanks for the information. Unfortunately I'm not able to contribute a PR because I don't know how `.frag` works in Angular apps. But I guess a quick check of `#version` or special types like `vec2`, `mat4` will work.

Actually, I think the classifier works quite well. The repo link I provided as a error case is from glslang and those files are test files, some of which contain syntax errors and some are actually hlsl. The rest are edge cases. Misclassification case is actually rare.

And my shader files are modified to be embedded as raw strings in c++ code, so probably that causes confusion to the classifier. I will fix my repo by appending `.glsl` to all shader files.

**Reply by zdzieblowski** [reply]
You can also add following override in your ``.gitattributes`` file:

```*.frag linguist-language=GLSL```
